far/vmenu.cpp

```diff
@@ -2011,79 +2097,115 @@ int VMenu::VisualPosToReal(int VPos) const
 
-	if (NewShowPos >= 0)
+	const auto NewHPos = [&]
 	{
-		Items[Item].ShowPos = std::min(static_cast<size_t>(NewShowPos), MaxShowPos);
-	}
-	else
-	{
-		Items[Item].ShowPos = MaxShowPos - std::min(static_cast<size_t>(-(NewShowPos + 1)), MaxShowPos);
-	}
+		if constexpr (requires { GetNewHPos(Item); })
+			return GetNewHPos(Item);
+		if constexpr (requires { GetNewHPos(ItemLength); })
+			return GetNewHPos(ItemLength);
+		if constexpr (requires { GetNewHPos(Item.HorizontalPosition, ItemLength); })
+			return GetNewHPos(Item.HorizontalPosition, ItemLength);
+	}();
 
-	if (Items[Item].ShowPos == OldShowPos)
-		return false;
+	UpdateAllItemsBoundaries(NewHPos, ItemLength);
+
+	if (Item.HorizontalPosition == NewHPos) return false;
 
-	VMFlags.Set(VMENU_UPDATEREQUIRED);
+	Item.HorizontalPosition = NewHPos;
 	return true;
 }
 
-bool VMenu::ShiftItemShowPos(int Item, int Shift, const size_t MaxLineWidth)
+bool VMenu::SetCurItemSmartHPos(const int NewHPos)
 {
-	const auto MaxShowPos{ GetItemMaxShowPos(Item, MaxLineWidth) };
-	auto ItemShowPos{ std::min(Items[Item].ShowPos, MaxShowPos) }; // Just in case
+	const auto TextAreaWidth{ CalculateTextAreaWidth() };
+	if (TextAreaWidth <= 0) return false;
+
+	const auto Policy{ CheckFlags(VMENU_ENABLEALIGNANNOTATIONS) ? item_hscroll_policy::cling_to_edge : item_hscroll_policy::bound_stick_to_left };
 
-	if (Shift >= 0)
+	if (SetItemHPos(
+		Items[SelectPos],
+		[&](const int ItemLength) { return get_item_smart_hpos(NewHPos, ItemLength, TextAreaWidth, Policy); }))
 	{
-		ItemShowPos += std::min(static_cast<size_t>(Shift), MaxShowPos - ItemShowPos);
+		SetMenuFlags(VMENU_UPDATEREQUIRED);
+		return true;
 	}
-	else
+
+	return false;
+}
+
+bool VMenu::ShiftCurItemHPos(const int Shift)
+{
+	const auto TextAreaWidth{ CalculateTextAreaWidth() };
+	if (TextAreaWidth <= 0) return false;
+
+	const auto Policy{ CheckFlags(VMENU_ENABLEALIGNANNOTATIONS) ? item_hscroll_policy::cling_to_edge : item_hscroll_policy::bound_stick_to_left };
+
+	if (SetItemHPos(
+		Items[SelectPos],
+		[&](const int ItemHPos, const int ItemLength) { return get_item_absolute_hpos(ItemHPos + Shift, ItemLength, TextAreaWidth, Policy); }))
 	{
-		ItemShowPos -= std::min(static_cast<size_t>(-Shift), ItemShowPos);
+		SetMenuFlags(VMENU_UPDATEREQUIRED);
+		return true;
 	}
 
-	if (ItemShowPos == Items[Item].ShowPos)
-		return false;
-
-	Items[Item].ShowPos = ItemShowPos;
-	VMFlags.Set(VMENU_UPDATEREQUIRED);
-	return true;
+	return false;
 }
 
-bool VMenu::SetAllItemsShowPos(int NewShowPos)
+template<typename GetNewHPos_T>
+bool VMenu::SetAllItemsHPos(GetNewHPos_T&& GetNewHPos)
 {
-	bool NeedRedraw = false;
-
-	const auto MaxLineWidth{ CalculateTextAreaWidth() };
+	ResetAllItemsBoundaries();
+	bool NeedRedraw{};
 
-	for (const auto I: std::views::iota(0uz, Items.size()))
-		NeedRedraw |= SetItemShowPos(static_cast<int>(I), NewShowPos, MaxLineWidth);
+	for (auto& Item : Items)
+		NeedRedraw |= SetItemHPos(Item, std::forward<GetNewHPos_T>(GetNewHPos));
```

Comment:
Sonarcloud [complains](https://sonarcloud.io/project/issues?resolved=false&pullRequest=789&id=FarGroup_FarManager&open=AY2Wvz1FTQdlhS0wyNVm) here, and technically quite rightfully:
1. By forwarding the argument we permit the callee to steal it.
2. The callee takes the parameter by value, so the parameter ctor technically can take advantage of that and steal from rvalue.
3. And then we pass this potentially empty thing again on the next iteration.

Given that everything here is local and we know exactly what comes in and don't need to store these callbacks or anything like that, I suggest keeping things simpler and passing callbacks by `const auto&` everywhere.

Reply:
Ouch! I've seen Sonarcloud's complaint but missed the fact that the forwarding happens in a loop. Thank you for checking.